client/lib/explat/internals/test/anon-id.ts

```diff
@@ -0,0 +1,175 @@
+		Array [
+		  Array [
+		    "Trying to initialize anonId outside of a browser context.",
+		  ],
+		]
+	` );
+	} );
+
+	it( 'should work as expected when the anonId is immediately available', async () => {
+		mockedGetTracksAnonymousUserId.mockImplementationOnce( () => 'anon-id-123' );
+		mockedGetTracksLoadPromise.mockImplementationOnce( () => Promise.resolve() );
+		expect( await AnonId.initializeAnonId() ).toBe( 'anon-id-123' );
+		expect( mockedRecordTracksEvent.mock.calls.length ).toBe( 1 );
+		expect( mockedGetTracksAnonymousUserId.mock.calls.length ).toBe( 1 );
+	} );
+
+	it( 'should work as expected when the anonId is available after a few tries', async () => {
+		jest.useFakeTimers();
+		mockedGetTracksLoadPromise.mockImplementationOnce( () => Promise.resolve() );
+		mockedGetTracksAnonymousUserId.mockImplementationOnce( () => null );
+		mockedGetTracksAnonymousUserId.mockImplementationOnce( () => undefined );
+		mockedGetTracksAnonymousUserId.mockImplementationOnce( () => '' );
+		mockedGetTracksAnonymousUserId.mockImplementationOnce( () => 'anon-id-123' );
+		const initializeAnonIdPromise = AnonId.initializeAnonId();
+		jest.runAllTimers();
+		expect( await initializeAnonIdPromise ).toBe( 'anon-id-123' );
+		expect( mockedRecordTracksEvent.mock.calls.length ).toBe( 1 );
+		expect( mockedGetTracksAnonymousUserId.mock.calls.length ).toBe( 4 );
+	} );
+
+	it( 'should poll at intervals', async () => {
+		jest.useFakeTimers();
+		mockedGetTracksLoadPromise.mockImplementationOnce( () => Promise.resolve() );
+		AnonId.initializeAnonId();
+		const intervalMs = 50;
+		jest.advanceTimersByTime( 0 );
+		expect( mockedGetTracksAnonymousUserId.mock.calls.length ).toBe( 1 );
+		jest.advanceTimersByTime( intervalMs - 1 );
+		expect( mockedGetTracksAnonymousUserId.mock.calls.length ).toBe( 1 );
+		jest.advanceTimersByTime( 1 );
+		expect( mockedGetTracksAnonymousUserId.mock.calls.length ).toBe( 2 );
+		jest.advanceTimersByTime( intervalMs - 1 );
+		expect( mockedGetTracksAnonymousUserId.mock.calls.length ).toBe( 2 );
+		jest.advanceTimersByTime( 1 );
+		expect( mockedGetTracksAnonymousUserId.mock.calls.length ).toBe( 3 );
+		jest.advanceTimersByTime( intervalMs - 1 );
+		expect( mockedGetTracksAnonymousUserId.mock.calls.length ).toBe( 3 );
+		jest.clearAllTimers();
+	} );
+
+	it( 'should give up after many polling attempts and return null', async () => {
+		jest.useFakeTimers();
+		mockedGetTracksLoadPromise.mockImplementationOnce( () => Promise.resolve() );
+		mockedGetTracksAnonymousUserId.mockImplementation( () => null );
+		const initializeAnonIdPromise = AnonId.initializeAnonId();
+		jest.runAllTimers();
+		expect( await initializeAnonIdPromise ).toBe( null );
+		expect( mockedRecordTracksEvent.mock.calls.length ).toBe( 1 );
+		expect( mockedGetTracksAnonymousUserId.mock.calls.length ).toBe( 100 );
+	} );
+
+	it( 'should give up immediately and return null if immediately logged in', async () => {
```

Comment:
Hmm... This may be a behaviour change from the current assumptions for experiments that cross the login boundary. However, I think it's covered by the assignment cookie, and in general we can't rely on the anon ID cookie being around forever. We should also be able to access the aliasing history on the backend if we move away from assignment cookies, so I think this is fine even if it's not the same as the existing client.

@withinboredom may have more thoughts.

Reply:
I think it won't hurt to do a quick check for anonId before we check for logged-in.